File: week3/sangmin/problem5/Problem5.java
package week3.sangmin.problem5;

import java.util.PriorityQueue;

class Solution {
    public int solution(int[] scoville, int K) {
        int answer = 0;
        PriorityQueue<Long> pq = new PriorityQueue<>();

        for(int s : scoville){
            pq.add((long) s);
        }

        while (pq.peek() < K && pq.size() >= 2){
            pq.add(pq.poll() + (2 * pq.poll()));
            answer++;
        }

        System.out.println(answer);

        if(answer == 0 || pq.peek() < K) return -1;
        return answer;
    }
}

public class Problem5 {
    public static void main(String[] args) {
        new Solution().solution(new int[]{1, 2, 3, 9, 10, 12}, 7);
        new Solution().solution(new int[]{1, 2}, 7);
    }
}

File: week3/inu/problem5/Problem5.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    priority_queue<int, vector<int>, greater<>> pq;
    for (int i : scoville)
        pq.push(i);

    while (true) {
        if (pq.size() < 2) {
            break;
        }

        if (pq.top() >= K)
            break;
        
        int a = pq.top(); pq.pop();
        int b = pq.top(); pq.pop();
        int c = a + (b * 2);
        pq.push(c);

        answer++;
    }

    if (pq.top() < K)
        answer = -1;

    return answer;
}

int main() {
    vector<int> scoville = { 1, 2, 3, 9, 10, 12 };
    int K = 100;
    cout << solution(scoville, K);
}